WordPress/openverse-frontend
# Convert `provider` store from Vuex to Pinia

## Fixes


Fixes #1027 by @obulat
Fixes #967 by @sarayourfriend 

**This PR was edited to convert the setup store to an options store to fix the server state hydration issues. Please, re-review :)**

## Description


This PR converts the `provider` store to Pinia. Provider stats include provider short name and display name, URL, as well as the media count per each provider. These data are used in the **provider filters for image/audio**, and on the **sources** page.

The state is simplified similar to the `media` store state to enable simple access to properties for various media types without string manipulation (e.g., `fetchState[mediaType].isFetching` instead of `isFetching${titleCase(mediaType)}Providers`).

Previously, we fetched provider data once, during the server initialization, using `nuxtServerInit` hook. Pinia does not have a direct analog.  Here, I used a `middleware` to call `fetchMediaProviders` before every route (found in [a Pinia discussion](https://github.com/vuejs/pinia/discussions/574)).
To minimize the number of unnecessary requests, I set a `lastUpdated` timestamp in the store as a `ssrRef`. When calling `fetchMediaProviders`, we only actually fetch data from the API if the provider array is empty, or more than 1 hour has passed since `lastUpdated`.
This way `provider` store loads all the provider stats for each request, and updates them if necessary once an hour. 

## Testing Instructions

Run the app, check the provider filters in the sidebar, they should work for images and audio.
Sources pages should display the image providers (note that audio providers were not yet added to this page).
Try disabling the network, you should see a warning in the console that providers were not loaded.

## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
